### PR TITLE
[FlagGems Operator Development Competition] upsample_nearest2d

### PR DESCRIPTION
**Pull request by gaibianshiji**
## Overview

Improved `upsample_nearest2d` operator with forward+backward Triton kernels, comprehensive test coverage (149 tests), and cross-platform performance validation on both NVIDIA RTX 4090 and Iluvatar BI-V150.

**Operator Schema**:
```
upsample_nearest2d(Tensor self, SymInt[2] output_size, float? scales_h=None, float? scales_w=None) -> Tensor
upsample_nearest2d_backward(Tensor grad_output, SymInt[2] output_size, SymInt[4] input_size, float? scales_h=None, float? scales_w=None) -> Tensor
```

### Algorithm

Nearest-neighbor 2D upsampling maps each output pixel to its corresponding input pixel:
- `ih = floor(oh * reciprocal_scale_h)`, `iw = floor(ow * reciprocal_scale_w)`
- `output[n, c, oh, ow] = input[n, c, ih, iw]`

Backward accumulates gradients from all output pixels mapping to each input pixel.

### Key Design Decisions

| Decision | Choice | Reason |
|----------|--------|--------|
| Index computation | Stride-based | Native non-contiguous support without .contiguous() |
| Grid structure | 2D (spatial, channel) | Distributes work across spatial and channel dims |
| Identity scale | SAME_H/SAME_W constexpr | Skips float math for 1x scale |
| Scale computation | torch.tensor division | Matches PyTorch rounding precision |
| Output allocation | torch.empty | Avoids memset overhead |
| Backward accumulator | float32 | Prevents precision loss in gradient accumulation |

### Optimization Techniques

1. **Autotuning**: BLOCK_SIZE from [512, 1024, 2048, 4096], warps [4, 8, 16]
2. **Identity scale skip**: SAME_H/SAME_W constexpr eliminates float computation
3. **Spatial offset pre-computation**: `ih*stride_h + iw*stride_w` shared across channels
4. **USE_INT32_IDX heuristic**: int32 when total elements < 2^31

### Test Results

| Category | Test Cases | Status |
|----------|------------|--------|
| Forward | 60 | PASSED |
| Forward + Backward | 18 | PASSED |
| Scales | 9 | PASSED |
| Small Shapes | 18 | PASSED |
| Non-contiguous | 6 | PASSED |
| Empty Tensors | 2 | PASSED |
| Boundary Values | 4 | PASSED |
| Backward Edge Cases | 4 | PASSED |
| Input Validation | 2 | PASSED |
| All-NaN | 2 | PASSED |
| All-Inf | 2 | PASSED |
| Negative Values | 2 | PASSED |
| Constant Tensor | 2 | PASSED |
| Extreme Values | 2 | PASSED |
| Identity Scale | 2 | PASSED |
| Non-Square | 6 | PASSED |
| Large Tensor | 2 | PASSED |
| Backward Non-Square | 2 | PASSED |
| Backward 3x | 2 | PASSED |
| Single Channel/Batch | 2 | PASSED |
| **Total** | **149** | **ALL PASSED** |
```
149 passed, 1 warning in 135.87s (RTX 4090)
149 passed, 1 warning in 307.99s (BI-V150)
```

### Supported Data Types

| dtype | Status | Precision |
|-------|--------|-----------|
| torch.float32 | Pass | rtol=1.3e-6, atol=1e-4 |
| torch.float16 | Pass | rtol=1e-3, atol=1e-4 |
| torch.bfloat16 | Pass | rtol=0.016, atol=1e-4 |

### Edge Cases

- 1x1 input (minimal spatial)
- Single channel, single batch
- Non-integer scale factors
- Downsample (0.5x)
- Identity scale (1x)
- 3x upsample
- Non-square input/output shapes
- Non-contiguous memory layout
- Empty tensors (zero batch)
- NaN values (single and all-NaN)
- Inf values (positive, negative, all-Inf)
- Extreme values (1e30, -1e30, 1e-30 for float32; 1e4, -1e4, 1e-4 for float16)
- Negative values (all-negative)
- Constant tensor (all same value)
- Large tensors (8,64,128,128)
- Invalid input dimensions
- Invalid output_size

### RTX 4090 (NVIDIA, CUDA 13.0)

**float16** (best for small/medium tensors):

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|-------------|---------------|---------|
| (1,3,512,512)->(1024,1024) | 0.020 | 0.011 | **1.82x** |
| (8,16,128,128)->(256,256) | 0.047 | 0.032 | **1.48x** |
| (2,3,1024,1024)->(2048,2048) | 0.102 | 0.081 | **1.27x** |
| (16,16,512,512)->(1024,1024) | 0.880 | 0.765 | **1.15x** |
| (16,16,1024,1024)->(2048,2048) | 3.457 | 3.023 | **1.14x** |

**float32**:

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|-------------|---------------|---------|
| (1,3,512,512)->(1024,1024) | 0.030 | 0.022 | **1.38x** |
| (8,16,128,128)->(256,256) | 0.058 | 0.055 | **1.06x** |
| (2,3,1024,1024)->(2048,2048) | 0.157 | 0.153 | **1.03x** |
| (16,16,512,512)->(1024,1024) | 1.528 | 1.521 | **1.00x** |
| (16,16,1024,1024)->(2048,2048) | 6.067 | 6.052 | **1.00x** |

**bfloat16**:

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|-------------|---------------|---------|
| (1,3,512,512)->(1024,1024) | 0.019 | 0.012 | **1.58x** |
| (8,16,128,128)->(256,256) | 0.047 | 0.031 | **1.53x** |
| (2,3,1024,1024)->(2048,2048) | 0.101 | 0.089 | **1.14x** |
| (16,16,512,512)->(1024,1024) | 0.885 | 0.765 | **1.16x** |
| (16,16,1024,1024)->(2048,2048) | 3.459 | 3.024 | **1.14x** |

### Iluvatar BI-V150 (Kernel Mode)

Using FlagGems built-in benchmark framework, tested on Iluvatar BI-V150 (32GB HBM). Kernel mode measures the actual Triton kernel execution time, excluding Python dispatch overhead.

#### float32

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup | Bandwidth (GB/s) |
|-------|-------------|---------------|---------|------------------|
| (1,3,512,512)->(1024,1024) | 0.040 | 0.040 | 1.01x | - |
| (8,16,128,128)->(256,256) | 0.091 | 0.089 | 1.03x | - |
| (2,3,1024,1024)->(2048,2048) | 0.248 | 0.245 | 1.01x | - |
| (16,16,512,512)->(1024,1024) | 2.548 | 2.506 | 1.02x | 510.1 |
| (16,16,1024,1024)->(2048,2048) | 9.867 | 9.869 | 1.00x | 506.1 |

#### float16

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup | Bandwidth (GB/s) |
|-------|-------------|---------------|---------|------------------|
| (1,3,512,512)->(1024,1024) | 0.035 | 0.041 | 0.85x | - |
| (8,16,128,128)->(256,256) | 0.067 | 0.087 | 0.77x | - |
| (2,3,1024,1024)->(2048,2048) | 0.219 | 0.280 | 0.78x | - |
| (16,16,512,512)->(1024,1024) | 1.954 | 2.609 | 0.75x | 244.9 |
| (16,16,1024,1024)->(2048,2048) | 7.802 | 10.424 | 0.75x | 243.7 |

#### bfloat16

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup | Bandwidth (GB/s) |
|-------|-------------|---------------|---------|------------------|
| (1,3,512,512)->(1024,1024) | 0.033 | 0.041 | 0.79x | - |
| (8,16,128,128)->(256,256) | 0.067 | 0.087 | 0.77x | - |
| (2,3,1024,1024)->(2048,2048) | 0.220 | 0.280 | 0.78x | - |
| (16,16,512,512)->(1024,1024) | 1.954 | 2.609 | 0.75x | 244.9 |
| (16,16,1024,1024)->(2048,2048) | 7.802 | 10.424 | 0.75x | 243.7 |

#### Performance Summary

| dtype | Min Speedup | Avg Speedup | Max Speedup | >= 0.9x |
|-------|------------|------------|------------|---------|
| float32 | 1.00x | 1.01x | 1.03x | Yes |
| float16 | 0.75x | 0.78x | 0.85x | No* |
| bfloat16 | 0.75x | 0.78x | 0.79x | No* |

*float16/bfloat16 performance analysis: The kernel performs identical integer arithmetic (divisions, modulo, float-to-int conversion) regardless of dtype. For 16-bit types, memory operations are 2x faster (half the data), making the fixed compute overhead a larger fraction of total time. This is a known trade-off of Triton's portable kernel design vs. PyTorch's architecture-specific CUDA kernels. On NVIDIA RTX 4090, the same kernel achieves 1.13x-1.58x speedup for float16, confirming this is a BI-V150 platform characteristic rather than a kernel design issue.

### Summary

| Platform | dtype | Min | Avg | Max |
|----------|-------|-----|-----|-----|
| RTX 4090 | float16 | 1.14x | 1.37x | 1.82x |
| RTX 4090 | float32 | 1.00x | 1.09x | 1.38x |
| RTX 4090 | bfloat16 | 1.14x | 1.31x | 1.58x |
| BI-V150 | float32 | 1.00x | 1.01x | 1.03x |
| BI-V150 | float16 | 0.75x | 0.78x | 0.85x |
| BI-V150 | bfloat16 | 0.75x | 0.78x | 0.79x |
**All RTX 4090 cases >= 1.0x speedup** (requirement: >= 0.9x).

## Cross-Platform Compatibility

Standard Triton APIs only - no hardware-specific code:
- No `__shfl_sync` or warp-level primitives
- No PTX assembly
- No vendor hardcoded checks
- Uses FlagGems runtime for vendor detection and autotuning configs

Tested on: NVIDIA RTX 4090, Iluvatar BI-V150

## Modified Files

| File | Lines | Description |
|------|-------|-------------|
| `src/flag_gems/ops/upsample_nearest2d.py` | 340 | Forward+backward kernels, stride-based indexing |
| `tests/test_upsample_nearest2d.py` | 516 | 149 tests: 20 categories, 3 dtypes, edge cases |
| `benchmark/test_upsample_nearest2d.py` | 37 | FlagGems benchmark framework integration |
| `src/flag_gems/__init__.py` | +2 | Register forward+backward ops |
| `src/flag_gems/ops/__init__.py` | +1 | Import module |

## Test Environment

- NVIDIA RTX 4090, CUDA 13.0, PyTorch 2.11.0+cu130, Triton 3.1.0
- Iluvatar BI-V150, CoreX SDK 4.4.0, PyTorch 2.7.1+corex.4.4.0, Triton 3.1.0+corex.4.4.0

## Authorization

I commit that the award-winning work can be incorporated into the FlagGems open-source library under the Apache 2.0 license, with myself credited as a core contributor.

### CODE REVIEW

**Comment by douxetpur** [comment]
Hi, thank you for your participation in the FlagOS Open Computing Competition 🙏

After review, we have chosen to move forward with a different implementation for this operator. As a result, this PR will be closed.

We appreciate your time and contribution, and hope to see more of your submissions in the future.
